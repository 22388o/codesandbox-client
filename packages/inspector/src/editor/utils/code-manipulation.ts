export enum InitializerType {
  String,
  Expression,
  Boolean,
}

function createProp(propName: string, initializerType: InitializerType) {
  const base = `${propName}`;

  switch (initializerType) {
    case InitializerType.Boolean:
      return base;
    case InitializerType.Expression:
      return `${base}={}`;
    case InitializerType.String:
      return `${base}=""`;
  }
}

/**
 * Returns the spaces (or tabs) the line starts with
 */
function getPadding(line: string | undefined): string {
  if (!line) {
    return '';
  }

  const matcher = /(\s*)/;
  const result = line.match(matcher);
  if (result && result[0]) {
    return result[0];
  } else {
    return '';
  }
}

export function addProp(
  oldCode: string,
  propName: string,
  initializerType: InitializerType
) {
  const lines = oldCode.split('\n');
  const lastLine = lines[lines.length - 1];
  const isSingleLine = lines.length === 1;
  const isSelfClosing = /\s?\/\s*>$/.test(lastLine);
  const ending = isSelfClosing ? /\s?\/\s*>/ : /\s?>/;
  const closeIsOwnLine = /^\s*\/?\s*>$/.test(lastLine);

  // Multiline component
  const padding = getPadding(lines[1]);

  if (closeIsOwnLine || !isSingleLine) {
    lines.splice(
      lines.length - 1,
      0,
      padding + createProp(propName, initializerType)
    );
  } else {
    lines[lines.length - 1] = lastLine.replace(
      ending,
      ' ' +
        createProp(propName, initializerType) +
        (isSelfClosing ? ' />' : '>')
    );
  }

  return lines.join('\n');
}
