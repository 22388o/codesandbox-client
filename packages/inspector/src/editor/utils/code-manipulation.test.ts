import { addProp, InitializerType } from './code-manipulation';

describe('single line', () => {
  it('can add a prop for self closing', () => {
    const singleLineComponent = `<Component a={5} />`;
    expect(
      addProp(singleLineComponent, 'test', InitializerType.Boolean)
    ).toMatchSnapshot();
    expect(
      addProp(singleLineComponent, 'test', InitializerType.Expression)
    ).toMatchSnapshot();
    expect(
      addProp(singleLineComponent, 'test', InitializerType.String)
    ).toMatchSnapshot();
  });

  it('can add a prop for open', () => {
    const singleLineComponent = `<Component a={5}>`;
    expect(
      addProp(singleLineComponent, 'test', InitializerType.Boolean)
    ).toMatchSnapshot();
    expect(
      addProp(singleLineComponent, 'test', InitializerType.Expression)
    ).toMatchSnapshot();
    expect(
      addProp(singleLineComponent, 'test', InitializerType.String)
    ).toMatchSnapshot();
  });
});

describe('multiline', () => {
  it('can add a prop', () => {
    const multiLineComponent = `<Component\n  a={5}\n/>`;
    expect(
      addProp(multiLineComponent, 'test', InitializerType.Boolean)
    ).toMatchSnapshot();
    expect(
      addProp(multiLineComponent, 'test', InitializerType.Expression)
    ).toMatchSnapshot();
    expect(
      addProp(multiLineComponent, 'test', InitializerType.String)
    ).toMatchSnapshot();
  });

  it('can add a prop to selfclose same line', () => {
    const multiLineComponent = `<Component\n  a={5}/>`;
    expect(
      addProp(multiLineComponent, 'test', InitializerType.Boolean)
    ).toMatchSnapshot();
    expect(
      addProp(multiLineComponent, 'test', InitializerType.Expression)
    ).toMatchSnapshot();
    expect(
      addProp(multiLineComponent, 'test', InitializerType.String)
    ).toMatchSnapshot();
  });
});
