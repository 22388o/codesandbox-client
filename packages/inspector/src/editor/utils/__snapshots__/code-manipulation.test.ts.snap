// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiline can add a prop 1`] = `
"<Component
  a={5}
  test
/>"
`;

exports[`multiline can add a prop 2`] = `
"<Component
  a={5}
  test={}
/>"
`;

exports[`multiline can add a prop 3`] = `
"<Component
  a={5}
  test=\\"\\"
/>"
`;

exports[`multiline can add a prop to selfclose same line 1`] = `
"<Component
  test
  a={5}/>"
`;

exports[`multiline can add a prop to selfclose same line 2`] = `
"<Component
  test={}
  a={5}/>"
`;

exports[`multiline can add a prop to selfclose same line 3`] = `
"<Component
  test=\\"\\"
  a={5}/>"
`;

exports[`single line can add a prop for open 1`] = `"<Component a={5} test>"`;

exports[`single line can add a prop for open 2`] = `"<Component a={5} test={}>"`;

exports[`single line can add a prop for open 3`] = `"<Component a={5} test=\\"\\">"`;

exports[`single line can add a prop for self closing 1`] = `"<Component a={5} test />"`;

exports[`single line can add a prop for self closing 2`] = `"<Component a={5} test={} />"`;

exports[`single line can add a prop for self closing 3`] = `"<Component a={5} test=\\"\\" />"`;
